core/storage/trie_db_overlay/impl/trie_db_overlay_impl.cpp

```diff
@@ -0,0 +1,160 @@
+      } else {
+        return trie::TrieError::NO_VALUE;
+      }
+    }
+    return storage_->get(key);
+  }
+
+  bool TrieDbOverlayImpl::contains(const Buffer &key) const {
+    auto it = extrinsics_changes_.find(key);
+    if (it != extrinsics_changes_.end()) {
+      return it->second.value.has_value();
+    }
+    return storage_->contains(key);
+  }
+
+  outcome::result<void> TrieDbOverlayImpl::put(const Buffer &key,
+                                               const Buffer &value) {
+    return put(key, Buffer{value});  // would have to copy anyway
+  }
+
+  outcome::result<void> TrieDbOverlayImpl::put(const Buffer &key,
+                                               Buffer &&value) {
+    auto extrinsic_id = getExtrinsicIndex();
+    auto it = extrinsics_changes_.find(key);
+    if (it != extrinsics_changes_.end()) {
+      it->second.value = std::move(value);
+      it->second.changers.emplace_back(extrinsic_id);
+      it->second.dirty = true;
+    } else {
+      extrinsics_changes_[key] = ChangedValue{
+          .value = std::move(value), .changers = {extrinsic_id}, .dirty = true};
+    }
+
+    return outcome::success();
+  }
+
+  outcome::result<void> TrieDbOverlayImpl::remove(const Buffer &key) {
+    auto extrinsic_id = getExtrinsicIndex();
+    auto it = extrinsics_changes_.find(key);
+    if (it != extrinsics_changes_.end()) {
+      it->second.value = boost::none;
+      it->second.dirty = true;
+      it->second.changers.emplace_back(extrinsic_id);
+    } else {
+      extrinsics_changes_[key] = ChangedValue{
+          .value = boost::none, .changers = {extrinsic_id}, .dirty = true};
+    }
+
+    return outcome::success();
+  }
+
+  outcome::result<void> TrieDbOverlayImpl::clearPrefix(
+      const common::Buffer &buf) {
+    // return outcome::result<void>();
+  }
+
+  Buffer TrieDbOverlayImpl::getRootHash() {
+    /// TODO(Harrm): process commit result; maybe not inherit overlay from trie
+    /// db after all?
+    commit();
+    return storage_->getRootHash();
+  }
+
+  bool TrieDbOverlayImpl::empty() const {
+    // gets a bit complex because some of the entries in changes are actually
+    // deleted entries
+    bool any_nondeleted_change =
+        std::any_of(
+        extrinsics_changes_.begin(),
+        extrinsics_changes_.end(), [](auto &change) {
+          return change.second.value.has_value();
+        });
+    /// TODO(harrm): check storage not cleared completely in cached chages
+    /// considering temporary values
+    return any_nondeleted_change or not storage_->empty();
+  }
+
+  primitives::ExtrinsicIndex TrieDbOverlayImpl::getExtrinsicIndex() const {
+    auto ext_idx_bytes_res = get(EXTRINSIC_INDEX_KEY);
+    if (ext_idx_bytes_res.has_value()) {
+      auto ext_idx_res =
+          scale::decode<primitives::ExtrinsicIndex>(ext_idx_bytes_res.value());
+      BOOST_ASSERT_MSG(ext_idx_res.has_value(),
+                       "Extrinsic index decoding must not fail");
```

Comment:
I would also print that in logger, so that if such issue happens on release build we can detect that